support path rules

nightlyone opened the pull request:
by allowing a placeholder.

explicit paths still have precedence.

This is useful in migration scenarios and when you don't want to deploy
just because someone added a repo to you organisation account.

nightlyone commented:
Hi @rakyll 

please take a look and give early feedback on the overall approach and what else you see.

Once it looks good, I will update the readme to reflect the new possibilities.

Many thanks :slightly_smiling_face:

nightlyone commented:
gentle ping @rakyll 

If you are in a hurry: To get a :+1: or :-1: on the basic idea without you reviewing everything,
having early feedback on https://github.com/GoogleCloudPlatform/govanityurls/pull/12/files#diff-c6c82e356db0ffdc1ce631b75c55e263R100 would be most useful to me, since those 25 lines it describe the intended new behavior best.

rakyll commented on handler.go:
```diff
@@ -37,20 +38,105 @@ type pathConfig struct {
+	if i == 0 {
+		 return "", "", "", fmt.Errorf("placeholder is empty in %q", match)
+	}
+	placeholder, s = s[:i], s[i+1:]
```
placeholder, suffix = s[:i], s[i+1:]

rakyll commented on handler.go:
```diff
@@ -37,20 +38,105 @@ type pathConfig struct {
+	}
+	placeholder, s = s[:i], s[i+1:]
+	suffix = s
+	if strings.ContainsAny(suffix, "{}") {
```
This is not going to match if there is anything between curly braces.

rakyll commented on handler.go:
```diff
@@ -26,8 +26,9 @@ import (
-	host  string
-	paths pathConfigSet
+	host      string
+	paths     pathConfigSet
```
why do we need both parsed paths and parsedPathRules? Can't we merge them into one type?

zombiezen replied:
:+1:

rakyll commented on handler.go:
```diff
@@ -37,20 +38,105 @@ type pathConfig struct {
-			Display string `yaml:"display,omitempty"`
-			VCS     string `yaml:"vcs,omitempty"`
-		} `yaml:"paths,omitempty"`
+type pathConfigRule struct {
```
These need to go into the pathConfig instead of introducing a new type.

rakyll commented on handler.go:
```diff
@@ -137,9 +254,13 @@ func (h *handler) Host(r *http.Request) string {
 {{range .Handlers}}<li><a href="https://godoc.org/{{.}}">{{.}}</a></li>{{end}}
 </ul>
+<ul>
+{{range .GenericRules}}<li>{{.Match}} will be clone repository {{.Subst}}></li>{{end}}
```
Just add the rule to the list above without a link and a placeholder.

nightlyone commented:
@rakyll fixed missing subpath feature and added a testcase for that. Addressed the smaller code issues.

About using only one struct to do both, I need a bit more time to think about how to do that. Will come back to you tomorrow with more ideas.

rakyll commented:
@nightlyone, thanks.

Also cc'ing @zombiezen for the review.

nightlyone commented:
Got lost here so closing to get it off everyones table.
